### PR TITLE
Add filter, virtualization, and export UI to DebugLogModal and unify file save

### PR DESCRIPTION
**Pull request by jschick04**
## Summary

Redesigns the debug log modal with filters, virtualized rendering, and export/copy actions, and unifies the existing file-save export pattern (used by 3 other modals) behind a new `IFileSaveService`.

> **Base branch:** targets `main`. Diff is rendered against `main`.

## What this changes

**DebugLogModal redesign**

- New `DebugLogEntry` record + `DebugLogEntryParser` parses `[ISO-Timestamp] [ThreadId] [Level] message` lines, folding continuation lines (stack traces) into the prior entry.
- New `DebugLogProjection` (pure) applies the active filters and reverses into newest-entry-first order via `ReversedListView<T>` (now implements `IList<T>` so Blazor `<Virtualize>` gets the `Enumerable.Skip` / `Take` indexer fast-path for deep offsets).
- Renders rows with `<Virtualize>` (`RowHeightPx=18` initial estimate, monospace font, `white-space: pre-wrap` + `overflow-wrap: anywhere` so long lines and stack traces wrap within the viewport instead of forcing a horizontal scrollbar). Multi-line entries render across multiple visual lines, which makes `<Virtualize>`''s scroll-position math approximate for tall rows — accepted tradeoff vs. the prior nowrap+ellipsis approach (which was inaccessible to keyboard / touch users) and the prior horizontal-scroll approach (which the user explicitly disliked).
- Filter bar adds a level filter (`ValueSelect` with `Equals` / `NotEqual` / `MultiSelect` operators) and a debounced (250ms) message string filter.
- Footer adds **Export to file**, **Copy to clipboard**, and a centered `X of Y entries` counter.
- `Refresh` wraps the streaming `LoadAsync` in `try/finally` so partial data and `_hasLoaded` remain consistent on exception (generation-guarded).
- `Clear` awaits `FileLogger.ClearAsync()` *first*, mutates UI state only on success, surfaces failures via `IAlertDialogService`.
- `DebugLogService.WriteTrace` formats the timestamp **inside** `_writeLock.EnterScope()` so the on-disk line order is monotonic in timestamp under contention. Backed by a new `Trace_WhenCalledFromMultipleThreadsConcurrently_*` regression test that fires 64×4 parallel writes and asserts (a) line count, (b) every line parses, (c) timestamps non-decreasing, (d) every unique payload appears exactly once.

**Unified file-save service**

- New `IFileSaveService` (in `EventLogExpert.UI.Interfaces`) + `FileSaveServiceFileTypes` (`Json` / `Log` typed dicts) as the single source of truth for file-type choices.
- New `MauiFileSaveService` implementation registered in `MauiProgram`. Adds `SetLength(0)` before `WriteAsync` so overwriting a larger existing file no longer leaves stale trailing bytes (rationale documented inline).
- Refactors **3 pre-existing export sites** (`FilterGroup.ExportGroup`, `FilterCacheModal.OnExportAsync`, `FilterGroupModal.OnExportAsync`) onto the new service, removing direct `Windows.Storage.Pickers` / `WinRT.Interop` dependencies from those `razor.cs` files.

## Tests

- 20 `DebugLogEntryParserTests` (prefix parsing, continuation folding, malformed lines, no-prior-entry edge cases).
- 21 `DebugLogProjectionTests` (level operators, text filter case-insensitivity, ordering, empty filters).
- 28 `DebugLogModalTests` (bUnit) covering filter bar interactions, counter text, export call, clear success/failure, copy.
- 6 new `ReversedListViewTests` covering `IndexOf` (reversed semantics), `Insert` / `RemoveAt` / indexer setter throwing `NotSupportedException`, and the `IList<T>` interface assertion.
- 1 new `DebugLogServiceTests` concurrent-write regression test (64×4 parallel writers, four invariants).
- All **1870** tests pass; build clean (0 warnings).

## Notes for reviewers

- **`MauiFileSaveService` truncation order** (`SetLength(0)` before `WriteAsync`) was deliberately chosen over a temp-file-and-rename pattern. Trade-off: a rare mid-write failure leaves a partial file on disk and surfaces the propagated exception (loudly visible) rather than the prior bug''s silent corruption (mix of new prefix + stale tail). Note this is not a temp-file-and-rename atomic write; partial-file-with-loud-failure is preferred over silent truncation-by-overwrite. Comment in source documents this.
- **`Clear()` reorder** is a small behavior change to a pre-existing path — file is now mutated before the UI, with explicit failure surfacing. Two new tests cover both branches.
- **Sibling alert message inconsistency** between `DebugLogModal` (`ex.Message`) and the 3 filter modals (`$"An exception occurred while exporting X: {ex.Message}"`) is intentional — each surface describes its own context.
- **Continuation-line ambiguity** in `DebugLogEntryParser` is documented and intentional: the parser classifies any line matching the `[ts] [tid] [Level] ` prefix as a new entry. A continuation line whose payload happens to start with that exact shape would be split as a new entry. Round 13 deliberately removed the previous timestamp-based heuristic because pre-PR-#513 writers formatted the timestamp **before** acquiring the lock, so persisted legacy `debug.log` files can legitimately contain headers in non-chronological order — a timestamp-based heuristic that folded an "older-looking" header into the prior entry would corrupt those legacy logs. The format is inherently ambiguous when message bodies contain header-shaped lines; we accept that ambiguity (no current producer in this repo intentionally emits header-shaped continuation content) rather than corrupt legacy logs.

### CODE REVIEW

File: src/EventLogExpert.Components/Modals/DebugLogModal.razor [hunk 1]
```diff
@@ -1,35 +1,106 @@
...
+@using Microsoft.Extensions.Logging
 
 <ModalChrome AriaLabel="Debug Log"
     CloseLabel="Exit"
     Footer="FooterPreset.CloseOnly"
     Height="60%"
     InlineAlert="@CurrentInlineAlert"
-    MinWidth="50rem"
+    MaxWidth="100rem"
+    MinWidth="100rem"
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.Components/Modals/DebugLogModal.razor:11]: Not changing this cause Virtualize rubber bands when scrolling due to different text lengths

File: src/EventLogExpert.UI/Services/DebugLogProjection.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        }
+    }
+
+    private static bool MatchesLevel(DebugLogEntry entry, FilterEvaluator op, IReadOnlyList<LogLevel> levels)
+    {
+        if (entry.Level is null) { return false; }
+
+        var contains = levels.Contains(entry.Level.Value);
+
+        return op == FilterEvaluator.NotEqual ? !contains : contains;
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.UI/Services/DebugLogProjection.cs:85]: Already fixed in this commit -- see `DebugLogProjection.cs:83` which returns `op == FilterEvaluator.NotEqual` for null Level (with new test `Project_WhenLevelNotEqualAndEntryHasNullLevel_ShouldIncludeEntry` in `DebugLogProjectionTests.cs`). Bot appears to have flagged stale data.

File: src/EventLogExpert.Components/Modals/DebugLogModal.razor [hunk 3]
```diff
@@ -1,35 +1,106 @@
...
-                    <br />
-                }
+                <ValueSelect AriaLabel="Levels"
+                    CssClass="input filter-dropdown debug-log-level-multi"
+                    IsMultiSelect
+                    T="LogLevel"
+                    Values="_multiLevels"
+                    ValuesChanged="HandleMultiLevelsChanged"
+                    ToStringFunc="@(x => x.ToString())">
+                    <ValueSelectItem T="LogLevel" ClearItem>All</ValueSelectItem>
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.Components/Modals/DebugLogModal.razor:39]: Already fixed in this commit -- see `DebugLogModal.razor:33` which sets `EmptyText="All"` on the multi-select. Bot appears to have flagged stale data.

File: src/EventLogExpert.Components/Modals/DebugLogModal.razor.cs [hunk 4]
```diff
@@ -35,52 +75,221 @@ protected override async Task OnInitializedAsync()
...
+
+    private void HandleMultiLevelsChanged(List<LogLevel> levels)
+    {
+        _multiLevels = levels;
+        ApplyProjection();
+    }
+
+    private void HandleSingleLevelChanged(LogLevel? level)
+    {
+        _singleLevel = level;
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.Components/Modals/DebugLogModal.razor.cs:202]: Fixed in 4af28a7. Refactored `FlushPendingStringFilter` into a paired `SyncPendingStringFilter` (state-only, returns bool) + `FlushPendingStringFilter` (sync + project). All three level handlers (`HandleLevelOperatorChanged`, `HandleMultiLevelsChanged`, `HandleSingleLevelChanged`) now call `SyncPendingStringFilter` before `ApplyProjection` so the just-typed string filter applies in the same single re-projection. New test `DebugLogModal_LevelFilterChangedWhilePendingFilterPending_FlushesPendingTextFilter` covers it.

File: src/EventLogExpert.Components/Modals/DebugLogModal.razor.cs [hunk 5]
```diff
@@ -35,52 +75,221 @@ protected override async Task OnInitializedAsync()
...
     {
         var generation = ++_loadGeneration;
 
         _hasLoaded = false;
-        _data.Clear();
+        var streamingEntries = new List<DebugLogEntry>();
+        _entries = streamingEntries;
+        SetDisplayed([]);
+        _filteredEntryCount = 0;
+        _projectedThroughIndex = 0;
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.Components/Modals/DebugLogModal.razor.cs:248]: Fixed in 4af28a7. `Refresh()` now calls `SyncPendingStringFilter()` at the top, so the streaming reload uses the textbox's current value rather than the stale `_activeStringFilter`. New test `DebugLogModal_RefreshClickedWhilePendingFilterPending_UsesNewFilterForReloadedEntries` covers it.

File: src/EventLogExpert.UI/Services/DebugLogEntryParser.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+            return false;
+        }
+
+        entry = new DebugLogEntry(timestamp, threadId, level, match.Groups["message"].Value, line);
+
+        return true;
+    }
+
+    [GeneratedRegex(@"^\[(?<ts>[^\]]+)\] \[(?<tid>\d+)\] \[(?<level>[A-Za-z]+)\] (?<message>.*)$")]
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.UI/Services/DebugLogEntryParser.cs:81]: Fixed in 4af28a7 with a monotonic-timestamp heuristic: when `TryParseLine` succeeds for a candidate header but its timestamp is older than the pending entry's timestamp, treat it as a continuation (since real log entries are written in chronological order). This catches the common case (logged exceptions that quote an older log line). New test `AddLine_WhenSubsequentHeaderHasOlderTimestamp_ShouldFoldAsContinuation` covers it. Note: the whitespace-prefix heuristic was considered but is moot — the regex's `^\[` anchor already rejects indented lines, so stack traces are safe today. The remaining unfixable case (an embedded excerpt with a *newer* timestamp than the pending entry) would require a format change to `DebugLogService.WriteTrace` to escape embedded `[...]` prefixes; documented in the `AddLine` xml-doc.

File: src/EventLogExpert.Components/Modals/DebugLogModal.razor.cs [hunk 7]
```diff
@@ -35,52 +75,235 @@ protected override async Task OnInitializedAsync()
...
+        if (_filteredEntryCount == 0) { return; }
+
+        try
+        {
+            await ClipboardService.CopyTextAsync(string.Join(Environment.NewLine, _displayedView));
+        }
+        catch (Exception ex)
+        {
+            await AlertDialogService.ShowAlert("Copy Failed", ex.Message, "OK");
+        }
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.Components/Modals/DebugLogModal.razor.cs]: Fixed in `db35bff`. Removed the dead `try`/`catch` from `HandleCopyAsync` and removed `DebugLogModal_CopyThrows_ShowsCopyFailedAlert` (it tested behavior that cannot occur in production). Also added XML doc comments on `IClipboardService` documenting the "best-effort, never throws" contract on the interface so future implementations and consumers understand the contract without re-deriving it from the implementation.

File: src/EventLogExpert.UI/Services/DebugLogEntryParser.cs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+    /// <summary>
+    ///     Feeds one line into the parser. Returns the previously buffered entry (with its continuation lines folded in)
+    ///     when <paramref name="line" /> starts a new entry; otherwise returns <c>null</c> and buffers the line internally.
+    /// </summary>
+    /// <remarks>
+    ///     A line whose prefix parses successfully is still treated as a continuation when its timestamp is older than the
+    ///     pending entry's timestamp, since real log entries are written in chronological order. This mitigates the case
+    ///     where a logged message contains a quoted excerpt of an older log line that happens to match the prefix shape.
+    /// </remarks>
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.UI/Services/DebugLogEntryParser.cs]: Fixed in `db35bff`. Moved the timestamp formatting and `Debug.WriteLine` inside `_writeLock.EnterScope()` in `DebugLogService.WriteTrace` so file order = chronological order under concurrent writes — the parser heuristic now operates on a stream where the monotonic-timestamp invariant actually holds. Confirmed via review that `EnsureWriter()` and the writer chain do not recursively call `WriteTrace`, so no deadlock risk against the `Lock` primitive.

File: src/EventLogExpert.UI/Interfaces/IFileSaveService.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    /// <summary>
+    ///     Opens a system "Save As" dialog and writes the provided text to the chosen file. Returns the saved path on
+    ///     success, or <c>null</c> if the user cancelled (or no host window is available). Throws if the user picked a path
+    ///     but the write failed; callers should handle failures via try/catch and surface them through
+    ///     <see cref="IAlertDialogService" />.
+    /// </summary>
+    /// <param name="suggestedFileName">Default filename shown in the dialog (e.g., "debug-log.log").</param>
+    /// <param name="fileTypes">Group label to extension list (e.g., "Log files" -> [".log", ".txt"]).</param>
+    /// <param name="content">UTF-8 text content to write.</param>
+    Task<string?> SaveTextAsync(
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.UI/Interfaces/IFileSaveService.cs]: Fixed in `db35bff`. `MauiFileSaveService` now throws `InvalidOperationException("No MAUI host window is available to present the Save As dialog.")` for the no-host case and reserves `null` strictly for genuine user cancellation. Updated the `IFileSaveService` XML doc to specify the new contract: returns `null` ONLY if the user cancelled, throws if the host environment cannot present the dialog or the underlying write fails. All four export call sites (DebugLogModal, FilterGroupModal, FilterCacheModal, FilterGroup) already wrap in `try`/`catch` with `IAlertDialogService.ShowAlert("Export Failed", ...)`, so the thrown exception now surfaces correctly to the user.

File: src/EventLogExpert.UI/Services/ReversedListView.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        ArgumentNullException.ThrowIfNull(inner);
+
+        _inner = inner;
+    }
+
+    public int Count => _inner.Count;
+
+    public bool IsReadOnly => true;
+
+    public T this[int index] => _inner[_inner.Count - 1 - index];
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.UI/Services/ReversedListView.cs]: Fixed in this round. `ReversedListView<T>` now implements `IList<T>` so `Enumerable.Skip(n).Take(m)` resolves to the `IListSkipTakeIterator<T>` indexer fast-path for the deep offsets `<Virtualize>` walks for each rendered window. `IndexOf` uses **reversed-view semantics** (linear walk via the projected indexer, not pass-through to `_inner.IndexOf` — pass-through would return inner-relative indices that consumers receive as projected-relative indices). Mutators (`Add`, `Insert`, `RemoveAt`, indexer setter, `Remove`, `Clear`) throw `NotSupportedException`; `IsReadOnly` returns `true` — the `IList<T> + NotSupportedException` shape that `Array.AsReadOnly` / `ReadOnlyCollection<T>` use. Backed by 6 new tests including `IndexOf_ShouldReturnPositionInReversedView`, `IndexOf_WhenItemAppearsTwice_ShouldReturnFirstOccurrenceInReversedView`, and `IsAssignableToIListT_ShouldExposeFastPathInterface` (pins the contract so a refactor cannot silently regress the fast path).

File: src/EventLogExpert.Components/Modals/DebugLogModal.razor [hunk 11]
```diff
@@ -1,35 +1,107 @@
...
+                @oninput="HandleStringFilterInput"
+                placeholder="Filter messages..."
+                type="text"
+                value="@_pendingStringFilter" />
+        </div>
+        <div aria-busy="@(_hasLoaded ? "false" : "true")" aria-label="Debug log entries" class="debug-log-viewport" role="region" tabindex="0">
+            <Virtualize Items="_displayedView" Context="line" ItemSize="@RowHeightPx" OverscanCount="20">
+                <ItemContent>
+                    <div class="debug-log-row">@line</div>
+                </ItemContent>
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.Components/Modals/DebugLogModal.razor:72]: Fixed in this round and PR description updated to match. The earlier `white-space: nowrap; text-overflow: ellipsis; + title="@line"` approach was reverted because (a) it relied on the `title` HTML attribute, which is mouse-hover-only and inaccessible to keyboard / touch users — i.e. there is no way for those users to see the truncated content, (b) `white-space: nowrap` collapses runs of internal whitespace per spec, which silently destroyed indentation in stack traces and aligned log columns, and (c) the user explicitly disliked the resulting horizontal scrollbar. The current CSS uses `white-space: pre-wrap` (preserves indentation, allows wrapping) + `overflow-wrap: anywhere` (long unbroken tokens like paths or base64 wrap instead of pushing horizontal overflow); the viewport sets `overflow-x: hidden` (explicit, not just absent) to forbid the scrollbar entirely. Tradeoff documented in PR description: variable-rendered-row-height under `<Virtualize ItemSize="18">` makes the scrollbar position approximate for tall rows. Per the user this is the preferred shape vs. the alternatives.

File: src/EventLogExpert.UI/Services/DebugLogEntryParser.cs [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        ArgumentNullException.ThrowIfNull(line);
+
+        if (DebugLogEntryParser.TryParseLine(line, out var parsed))
+        {
+            var emitted = BuildPending();
+
+            _pending = parsed;
+            _continuationLines = null;
+
+            return emitted;
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.UI/Services/DebugLogEntryParser.cs:120]: Acknowledged tradeoff and intentional. Two reasons round 13 deliberately accepts this ambiguity:

1. **No current in-repo producer intentionally emits header-shaped continuation content.** The single multi-line writer is `OnUnhandledException` formatting `Exception.ToString()`, which produces stack-trace lines starting with `   at ` (leading whitespace, not a `[`). Other multi-line content (rare) is exception messages whose bodies are not under our control; if those happen to contain a `[ts] [tid] [Level] ` prefix verbatim, the split is the documented behavior, not a bug.

2. **The previous timestamp-based heuristic actively corrupted legacy logs.** Pre-PR-#513 writers formatted the timestamp **before** acquiring `_writeLock`, so persisted `debug.log` files on disk can legitimately contain headers whose timestamps are out of order (Thread A captures `T=10ms`, gets descheduled, Thread B captures `T=11ms` and writes first, A then writes `T=10ms` after — file now has `[T=11ms]...\n[T=10ms]...` as a valid sequence). A heuristic that folds the second header into the previous entry "because its timestamp went backwards" deletes a valid entry. Round 13 removed that heuristic and round 14 added a `Trace_WhenCalledFromMultipleThreadsConcurrently_*` regression test that pins the producer-side fix (timestamp now formatted inside the lock) so the on-disk invariant holds going forward — but the parser still has to read legacy files written before the producer fix landed, so it cannot rely on the invariant retroactively.

Stronger producer-side defenses (escape `[` in continuation content, length-prefixed framing, explicit `---ENTRY---` separator) are listed under the "Stateful parsers ambiguous when continuation matches header shape" rule in the personal instructions file as the structural fix; deferring as a follow-up to keep this PR scoped to the perf and UX fixes the user asked for.

Resolving as wont-fix.

File: src/EventLogExpert.UI/Services/DebugLogService.cs [hunk 13]
```diff
@@ -181,16 +181,16 @@ private void OnUnhandledException(object sender, UnhandledExceptionEventArgs e)
...
         {
+            string output = $"[{DateTime.Now:o}] [{Environment.CurrentManagedThreadId}] [{level}] {message}";
+
             EnsureWriter();
             _writer?.WriteLine(output);
-        }
 
 #if DEBUG
-        System.Diagnostics.Debug.WriteLine(output);
+            System.Diagnostics.Debug.WriteLine(output);
```
[the remark this answers is not in the document]
Comment (jschick04) [reply src/EventLogExpert.UI/Services/DebugLogService.cs:222]: Fixed in this round. New `Trace_WhenCalledFromMultipleThreadsConcurrently_ShouldProduceOrderedAndCompleteOutput` in `DebugLogServiceTests` fires 64 writers × 4 writes each (256 total) with a `TaskCompletionSource` start gate so all writers race for the lock, then awaits all and asserts ALL of:

1. **Line count == 256** — no writes lost or duplicated at the file level.
2. **Every line parses** as `[ts] [tid] [Level] message` via the same regex `DebugLogEntryParser` uses — no interleaved/garbled prefixes.
3. **Timestamps non-decreasing** (`<=`, not `<` — `DateTime.Now` resolution can produce equal captures inside the lock and that is valid; what we forbid is a regression where a later line in the file has an earlier timestamp than the previous line).
4. **Every unique payload appears exactly once** via HashSet equality between `expected = {writer-{w}-msg-{i} | w∈[0,64), i∈[0,4)}` and the parsed `payloads.ToHashSet()`. Count-only assertions silently pass on duplicate-and-drop bugs; HashSet equality catches them.

Without the per-payload uniqueness assertion, a "wrote N items in order" check could not distinguish "all 256 unique payloads written" from "wrote 256 items but lost one and duplicated another." A future change that moves the timestamp capture back outside the lock would fail invariant (3) under contention; a change that loses or duplicates writes would fail (1) or (4).

**Review by NikTilton — APPROVED** [review]
